README.md
# Hexdump
[![Build Status](https://travis-ci.org/zuckel/hexdump.svg?branch=master)](https://travis-ci.org/zuckel/hexdump)
[![codecov](https://codecov.io/gh/zuckel/hexdump/branch/master/graph/badge.svg)](https://codecov.io/gh/zuckel/hexdump)
[![Latest release](https://img.shields.io/github/release/zuckel/hexdump.svg)](https://github.com/zuckel/hexdump/releases/latest)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.lasinger.tools/hexdump/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.lasinger.tools/hexdump)

This repo implements a simple hexdump, similar in output to `hexdump -C` of BSD hexdump. It will transform a byte[] to a String.

## GAV
```
org.lasinger.tools:hexdump:VERSION
```

## Usage
The interface is a single static method in `org.lasinger.tools.hexdump.Hexdump`.

```
public static String hexdump(byte[])
```

## Output
There are minor deviations from BSD hexdump:
- ascii control bytes TAB, LF, CR, SP are output as readable unicode characters ( → ¶ ¤ ␣ )
- other non-printable ascii bytes are output as WHITE SQUARE ( □ )

```
00000000  00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f  |□□□□□□□□□→¶□□¤□□|
00000010  10 11 12 13 14 15 16 17  18 19 1a 1b 1c 1d 1e 1f  |□□□□□□□□□□□□□□□□|
00000020  20 21 22 23 24 25 26 27  28 29 2a 2b 2c 2d 2e 2f  |␣!"#$%&'()*+,-./|
00000030  30 31 32 33 34 35 36 37  38 39 3a 3b 3c 3d 3e 3f  |0123456789:;<=>?|
00000040  40 41 42 43 44 45 46 47  48 49 4a 4b 4c 4d 4e 4f  |@ABCDEFGHIJKLMNO|
00000050  50 51 52 53 54 55 56 57  58 59 5a 5b 5c 5d 5e 5f  |PQRSTUVWXYZ[\]^_|
00000060  60 61 62 63 64 65 66 67  68 69 6a 6b 6c 6d 6e 6f  |`abcdefghijklmno|
00000070  70 71 72 73 74 75 76 77  78 79 7a 7b 7c 7d 7e 7f  |pqrstuvwxyz{|}~□|
00000080  80 81 82 83 84 85 86 87  88 89 8a 8b 8c 8d 8e 8f  |□□□□□□□□□□□□□□□□|
00000090  90 91 92 93 94 95 96 97  98 99 9a 9b 9c 9d 9e 9f  |□□□□□□□□□□□□□□□□|
000000a0  a0 a1 a2 a3 a4 a5 a6 a7  a8 a9 aa ab ac ad ae af  |□□□□□□□□□□□□□□□□|
000000b0  b0 b1 b2 b3 b4 b5 b6 b7  b8 b9 ba bb bc bd be bf  |□□□□□□□□□□□□□□□□|
000000c0  c0 c1 c2 c3 c4 c5 c6 c7  c8 c9 ca cb cc cd ce cf  |□□□□□□□□□□□□□□□□|
000000d0  d0 d1 d2 d3 d4 d5 d6 d7  d8 d9 da db dc dd de df  |□□□□□□□□□□□□□□□□|
000000e0  e0 e1 e2 e3 e4 e5 e6 e7  e8 e9 ea eb ec ed ee ef  |□□□□□□□□□□□□□□□□|
000000f0  f0 f1 f2 f3 f4 f5 f6 f7  f8 f9 fa fb fc fd fe ff  |□□□□□□□□□□□□□□□□|
00000100


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2019-01-22
### Changed
- Return "empty" on empty input and "null" on null input to make all output human-readable.
- Change non-printable byte representation for better distinction from dot.

## [0.1.0] - 2018-12-30
### Added
- Initial release
- Added Hexdump.hexdump(byte[])


src/test/java/org/lasinger/tools/hexdump/HexdumpTest.java
package org.lasinger.tools.hexdump;

import static java.nio.charset.StandardCharsets.UTF_8;
import static net.trajano.commons.testing.UtilityClassTestUtil.assertUtilityClassWellDefined;
import static org.assertj.core.api.Assertions.assertThat;
import static org.lasinger.tools.hexdump.Hexdump.hexdump;

import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.Test;

public class HexdumpTest {

    @Test
    public void utilityClass() throws Exception {
        assertUtilityClassWellDefined(Hexdump.class);
    }

    @Test
    public void example_gitignore() throws Exception {
        String in = "/.classpath\n" + //
                "/.project\n" + //
                "/.settings/\n" + //
                "/target/\n" + //
                "/nbproject/\n" + //
                "/nb-configuration.xml";
        System.out.println(hexdump(in.getBytes(UTF_8)));
    }

    @Test
    public void nullBytes() throws Exception {
        assertThat(hexdump(null)).isEqualTo("null");
    }

    @Test
    public void noBytes() throws Exception {
        assertThat(hexdump(new byte[0])).isEqualTo("empty");
    }

    @Test
    public void hexEncoding() throws Exception {
        String in = "1fe20bc4";
        int i = Integer.parseInt(in, 16);
        byte[] bytes = new byte[] { (byte) (i >>> 24), (byte) (i >>> 16), (byte) (i >>> 8), (byte) i };

        String dump = hexdump(bytes);

        System.out.println(dump);
        assertThat(dump.replaceAll(" ", "")).contains(in);
    }

    @Test
    public void eightBytes() throws Exception {
        String dump = hexdump(new byte[] { 48, 49, 50, 51, 52, 53, 54, 55 });

        System.out.println(dump);
        assertThat(dump).startsWith("00000000  30 31 32 33 34 35 36 37                           |01234567|");
    }

    @Test
    public void nineBytes() throws Exception {
        String dump = hexdump(new byte[] { 48, 49, 50, 51, 52, 53, 54, 55, 56 });

        System.out.println(dump);
        assertThat(dump).startsWith("00000000  30 31 32 33 34 35 36 37  38                       |012345678|");
    }

    @Test
    public void specialBytes() throws Exception {
        String dump = hexdump(new byte[] { 0x09, 0x20, 0x20, 0x0D, 0x0A, 0x2E, 0x00 });

        System.out.println(dump);
        assertThat(dump).startsWith("00000000  09 20 20 0d 0a 2e 00                              |→␣␣¤¶.□|");
    }

    @Test
    public void edgeBytes() throws Exception {
        String dump = hexdump(new byte[] { 0x00, Byte.MIN_VALUE, Byte.MAX_VALUE, (byte) 255 });

        System.out.println(dump);
        assertThat(dump).startsWith("00000000  00 80 7f ff                                       |□□□□|");
    }

    @Test
    public void singleLine() throws Exception {
        String dump = hexdump("1234".getBytes(UTF_8));
        assertThat(dump).isEqualTo("00000000  31 32 33 34                                       |1234|\n" //
                + "00000004\n");
    }

    @Test
    public void singleLineFull() throws Exception {
        String dump = hexdump("1234567812345678".getBytes(UTF_8));
        assertThat(dump).isEqualTo("00000000  31 32 33 34 35 36 37 38  31 32 33 34 35 36 37 38  |1234567812345678|\n" //
                + "00000010\n");
    }

    @Test
    public void multiLine() throws Exception {
        String dump = hexdump("1234567812345678abc".getBytes(UTF_8));
        assertThat(dump).isEqualTo("00000000  31 32 33 34 35 36 37 38  31 32 33 34 35 36 37 38  |1234567812345678|\n"
                + "00000010  61 62 63                                          |abc|\n" //
                + "00000013\n");
    }

    @Test
    public void printableChars() throws Exception {
        String nonAscii = "äöüß";

        String in = " !\"#$%&'()*+,-./0123456789:;<=>?@ABC-" + nonAscii + "-XYZ[\\]^_`abc-xyz{|}~";

        String dump = hexdump(in.getBytes(UTF_8));

        System.out.println(dump);
        String printableAscii = in.replace(nonAscii, "");
        assertThat(codepointList(dump)).containsAll(codepointList(printableAscii));
        assertThat(codepointList(dump)).doesNotContainAnyElementsOf(codepointList(nonAscii));
        assertThat(dump).contains("ABC-□□□□□□□□-X");
    }

    // this test currently has no assertion. it is for visual output only.
    @Test
    public void fullByteRange() throws Exception {
        byte[] allBytes = new byte[256];
        for (int i = 0; i < 256; i++) {
            allBytes[i] = (byte) i;
        }
        System.out.println(hexdump(allBytes));
    }

    private List<Integer> codepointList(String string) {
        return string.codePoints().collect(ArrayList::new, List::add, List::addAll);
    }

}


src/main/java/org/lasinger/tools/hexdump/Hexdump.java
package org.lasinger.tools.hexdump;

/**
 * Utility class for generating hexdumps from byte arrays. Mostly for debugging purposes.
 */
public final class Hexdump {

    private static final char[] HEX =
            new char[] { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };

    public static final char NON_PRINTABLE = '\u25a1'; // WHITE SQUARE □

    private Hexdump() {
    }

    /**
     * <p>
     * Create human-readable hexdump for a byte array.
     * </p>
     * <p>
     * This method is not thread-safe in the sense that bytes will be read more than once, thus possibly producing inconsistent
     * output if the byte array is mutated concurrently.
     * </p>
     *
     * @param bytes array to be rendered
     * @return human-readable formatted hexdump as a String
     */
    public static String hexdump(byte[] bytes) {
        if (bytes == null) {
            return "null";
        }
        if (bytes.length == 0) {
            return "empty";
        }

        StringBuilder out = new StringBuilder();

        for (int offset = 0; offset < bytes.length; offset += 16) {
            appendLine(bytes, offset, out);
        }
        appendOffset(bytes.length, out);
        out.append('\n');

        return out.toString();
    }

    private static void appendLine(byte[] bytes, int firstBlockStart, StringBuilder out) {
        int firstBlockEnd = Math.min(bytes.length, firstBlockStart + 8);
        int secondBlockEnd = Math.min(bytes.length, firstBlockStart + 16);

        appendOffset(firstBlockStart, out);
        out.append(' ');
        out.append(' ');

        appendHexBytes(bytes, firstBlockStart, firstBlockEnd, out);
        out.append(' ');

        appendHexBytes(bytes, firstBlockStart + 8, secondBlockEnd, out);
        padMissingBytes(firstBlockStart, secondBlockEnd, out);
        out.append(' ');

        out.append('|');
        appendDisplayChars(bytes, firstBlockStart, secondBlockEnd, out);
        out.append('|');
        out.append('\n');
    }

    private static void appendOffset(int offset, StringBuilder out) {
        int x = offset;
        for (int i = 0; i < 4; i++) {
            x = Integer.rotateLeft(x, 8);
            Hexdump.appendHexChars((byte) (x & 0xFF), out);
        }
    }

    private static void appendHexBytes(byte[] bytes, int offset, int limit, StringBuilder out) {
        for (int i = offset; i < limit; i++) {
            appendHexChars(bytes[i], out);
            out.append(' ');
        }
    }

    private static void appendHexChars(byte b, StringBuilder out) {
        out.append(HEX[(b >> 4) & 0x0F]); // 4 high bits
        out.append(HEX[b & 0x0F]); // 4 low bits
    }

    private static void padMissingBytes(int firstByte, int lastByte, StringBuilder out) {
        int charsPerByte = 3;
        int maxBytesPerLine = 16;
        int bytesWritten = lastByte - firstByte;

        int charsMissing = charsPerByte * (maxBytesPerLine - bytesWritten);
        for (int i = 0; i < charsMissing; i++) {
            out.append(' ');
        }
    }

    private static void appendDisplayChars(byte[] bytes, int offset, int blockEnd, StringBuilder out) {
        for (int i = offset; i < blockEnd; i++) {
            appendDisplayChar(bytes[i], out);
        }
    }

    private static void appendDisplayChar(byte b, StringBuilder out) {
        switch (b) {
        case 0x20:
            out.append("\u2423"); // SPACE ␣
            break;
        case 0x09:
            out.append('\u2192'); // TAB →
            break;
        case 0x0a:
            out.append('\u00b6'); // LF ¶
            break;
        case 0x0d:
            out.append('\u00a4'); // CR ¤
            break;
        default:
            out.append((32 <= b && b <= 126) ? (char) b : NON_PRINTABLE); // ' ' to '~', non-printable is WHITE SQUARE □
        }

    }

}
